File: high/025/solution.js
// ===== 25. K 个一组翻转链表 =====
// 难度: 困难
// 英文名: Reverse Nodes in k-Group
// 来源: https://leetcode.cn/problems/reverse-nodes-in-k-group/description/
//
// 给你链表的头节点 head ，每 k 个节点一组进行翻转，请你返回修改后的链表。

// k 是一个正整数，它的值小于或等于链表的长度。如果节点总数不是 k 的整数倍，那么请将最后剩余的节点保持原有顺序。

// 你不能只是单纯的改变节点内部的值，而是需要实际进行节点交换。

// 示例 1：

// 输入：head = [1,2,3,4,5], k = 2
// 输出：[2,1,4,3,5]
// 示例 2：

// 输入：head = [1,2,3,4,5], k = 3
// 输出：[3,2,1,4,5]

// 提示：
// 链表中的节点数目为 n
// 1 <= k <= n <= 5000
// 0 <= Node.val <= 1000

// 进阶：你可以设计一个只用 O(1) 额外内存空间的算法解决此问题吗？
// ---------------------------------------------------------

function ListNode(val, next) {
  this.val = val;
  this.next = next ?? null;
}

function reverseKGroup(head, k) {
  let n = 0;
  for (let cur = head; cur; cur = cur.next) n++;

  const dummy = new ListNode(0, head);

  let p0 = dummy;
  let pre = null;
  let cur = head;

  for (; n >= k; n -= k) {
    for (let i = 0; i < k; i++) {
      const next = cur.next;
      cur.next = pre;
      pre = cur;
      cur = next;
    }

    const next = p0.next;
    p0.next.next = cur;
    p0.next = pre;
    p0 = next;
  }
  return dummy.next;
}

function arrayToList(arr) {
  const dummy = new ListNode(0);
  let cur = dummy;
  for (const val of arr) {
    cur.next = new ListNode(val);
    cur = cur.next;
  }
  return dummy.next;
}

function listToArray(head) {
  const result = [];
  while (head) {
    result.push(head.val);
    head = head.next;
  }
  return result;
}

console.log(listToArray(reverseKGroup(arrayToList([1, 2, 3, 4, 5]), 2))); // [2,1,4,3,5]
console.log(listToArray(reverseKGroup(arrayToList([1, 2, 3, 4, 5]), 3))); // [3,2,1,4,5]

File: high/025/solution.py
# ===== 25. K 个一组翻转链表 =====
# 难度: 困难
# 英文名: Reverse Nodes in k-Group
# 来源: https://leetcode.cn/problems/reverse-nodes-in-k-group/description/
#
# https://leetcode.com/problems/reverse-nodes-in-k-group/
#
# Given a linked list, reverse the nodes of a linked list k at a time and return its modified list.
#
# k is a positive integer and is less than or equal to the length of the linked list. If the number of nodes is not a multiple of k then left-out nodes in the end should remain as it is.
#
# Example:
#
# Given this linked list: 1->2->3->4->5
#
# For k = 2, you should return: 2->1->4->3->5
#
# For k = 3, you should return: 3->2->1->4->5
#
# Note:
#
# Only constant extra memory is allowed.
# You may not alter the values in the list's nodes, only nodes itself may be changed.
#
# ---------------------------------------------------------


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# def reverseKGroup(head, k):
#     """反转链表每 k 个节点一组"""
#     # 统计链表长度
#     n = 0
#     cur = head
#     while cur:
#         n += 1
#         cur = cur.next

#     dummy = ListNode(0, head)
#     p0 = dummy
#     cur = head

#     while n >= k:
#         pre = None

#         # 反转 k 个节点
#         for _ in range(k):
#             nxt = cur.next
#             cur.next = pre
#             pre = cur
#             cur = nxt

#         # 连接反转后的链表
#         tail = p0.next
#         tail.next = cur
#         p0.next = pre
#         p0 = tail

#         n -= k

#     return dummy.next


def reverseKGroup(head, k):
    n = 0
    cur = head
    while cur:
        n += 1
        cur = cur.next
    dummy = ListNode(0, head)
    p0 = dummy
    cur = head
    while n >= k:
        pre = None

        for _ in range(k):
            nxt = cur.next
            cur.next = pre
            pre = cur
            cur = nxt
        tail = p0.next
        tail.next = cur
        p0.next = pre
        p0 = tail
        n -= k
    return dummy.next


def arrayToList(arr):
    dummy = ListNode()
    cur = dummy
    for val in arr:
        cur.next = ListNode(val)
        cur = cur.next
    return dummy.next


def listToArray(head):
    result = []
    while head:
        result.append(head.val)
        head = head.next
    return result


# 测试
print(listToArray(reverseKGroup(arrayToList([1, 2, 3, 4, 5]), 2)))  # [2, 1, 4, 3, 5]
print(listToArray(reverseKGroup(arrayToList([1, 2, 3, 4, 5]), 3)))  # [3, 2, 1, 4, 5]
print(listToArray(reverseKGroup(arrayToList([1, 2]), 3)))  # [1, 2]
